kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py

```diff
@@ -0,0 +1,220 @@
+
+class KFHungarianTracker(Node):
+    '''Use Kalman Fiter and Hungarian algorithm to track multiple dynamic obstacles
+
+    Use Hungarian algorithm to match presenting obstacles with new detection and maintain a kalman filter for each obstacle.
+    spawn ObstacleClass when new obstacles come and delete when they disappear for certain number of frames
+
+    Attributes:
+        obstacle_list: a list of ObstacleClass that currently present in the scene
+        max_id: the maximum id assigned 
+        sec, nanosec: timing from sensor msg
+        detection_sub: subscrib detection result from detection node
+        tracker_obstacle_pub: publish tracking obstacles with ObstacleArray
+        tracker_pose_pub: publish tracking obstacles with PoseArray, for rviz visualization
+    '''
+
+    def __init__(self):
+        '''initialize attributes and setup subscriber and publisher'''
+
+        super().__init__('kf_hungarian_node')
+        self.declare_parameters(
+            namespace='',
+            parameters=[
+                ('a_noise', [2., 2., 0.5]),
+                ('top_down', False),
+                ('death_threshold', 3),
+                ('measurementNoiseCov', [1., 1., 1.]),
+                ('errorCovPost', [1., 1., 1., 10., 10., 10.]),
+                ('vel_filter', 0.2),
+                ('cost_filter', 1.0)
+            ])
+        self.death_threshold = self.get_parameter("death_threshold")._value
+        self.measurementNoiseCov = self.get_parameter("measurementNoiseCov")._value
+        self.errorCovPost = self.get_parameter("errorCovPost")._value
+        self.a_noise = self.get_parameter("a_noise")._value
+        self.vel_filter = self.get_parameter("vel_filter")._value
+        self.top_down = self.get_parameter("top_down")._value
+        self.cost_filter = self.get_parameter("cost_filter")._value
+
+        self.obstacle_list = []
+        self.max_id = 0
+        self.sec = 0
+        self.nanosec = 0
+
+        # subscribe to detector 
+        self.detection_sub = self.create_subscription(
+            ObstacleArray,
+            "detection",
+            self.callback,
+            10)
+
+        # publisher for tracking result
+        self.tracker_obstacle_pub = self.create_publisher(ObstacleArray, 'tracking', 10)
+        self.tracker_marker_pub = self.create_publisher(MarkerArray, 'marker', 10)
+
+    def callback(self, msg):
+        '''callback function for detection result'''
+
+        # update delta time
+        dt = (msg.header.stamp.sec - self.sec) + (msg.header.stamp.nanosec - self.nanosec) / 1e9
+        self.sec = msg.header.stamp.sec
+        self.nanosec = msg.header.stamp.nanosec
+
+        # get detection
+        detections = msg.obstacles
+        num_of_detect = len(detections)
+        num_of_obstacle = len(self.obstacle_list)
+
+        # kalman predict
+        for obj in self.obstacle_list:
+            obj.predict(dt)
+
+        # hungarian matching
```

Comment:
Before this point, we need to transform the `msg`'s new observations into a `global_frame` parameter frame. That way we track while a robot / camera is in motion. I think it would be reasonable to have `global_frame` default to `None` and if `None` then use the `msg`'s frame (e.g. don't do anything).